### PR TITLE
✨ [amp-consent] Add CookieWriter feature to sync CID from amp-analytics LINKER

### LINKED ISSUE #30566: Set CID Cookie before promptUiSrc loads
## What's the issue?
We are using google's CLIENT_ID to sync up with our internal consent uuids to make sure we can properly persist the users consent preferences in the absence of cookies. I seem to be running into a problem with CLIENT_ID being different when the amp page is hosted off the publisher website vs. off the google.com TLD.

## How do we reproduce the issue?
we pass the CLIENT_ID macro in the promptUISrc url

the results:
e.g. 
https://www.google.com/amp/s/www.publisher.com/amp/en-us/content-video-123

has a client id of something like: SPB8rlQkK8dFUnpCGjto31NL7P5g-j06ZJAn8tOoWv3G4oZdpw-deDoCn9YTiqhJ

but when you go to the publishers amp page of that same content in that same browser without clearing any cookies/local storage: 
https://www.publisher.com/amp/en-us/content-video-123
the client id is something like amp-nOpe-wXCUM1STTYOlYuSqg

i think the expected behavior would be that the client id's would be the same, so we can properly sync up the client id with our user consent uuid. but because it is different, the consent preferences of the user does not persist. 

What is the best way to resolve this? because when amp pages are served off the google.com TLD, the cookies arent accessible when they go to a publishers page, so we are relying on google to provide a consistent CLIENT_ID

## What browsers are affected?
all

## Which AMP version is affected?
all AFAIK

_Originally posted by @vyoungnyc in https://github.com/ampproject/amphtml/issues/26229#issuecomment-698679599_

### PR DESCRIPTION
**Pull request by micajuine-ho**
Closes #30566

**Overview**
`amp-analytics` uses the `Linker` feature and the `CookieWriter` feature to sync CIDs from cache to origin, when a user navigates from cache to origin. `amp-consent` sends CID to CMP/developer endpoints to sync consent state for user on both cache and origin. Ideally, when the user navigates to the origin, the CID passed from the Linker would be written into cookies by `amp-analytics` and then `amp-consent` would send the sync'd CID to the CMP/developer consent endpoint. However, this a race, since general analytics loads after consent.

**Solution**
We shouldn't make CID generation wait for the Linker to finish, as it would very negatively affect critical loading path for users. So therefore, we need to parse the Linker CID before we hit the CMP/developer endpoint.

In the end, we would like to split the Linker/CookieWriter features from `amp-analytics` (and `amp-consent`) turn it into an extension or service, that can be loaded (lazily or even added by the cache). This way both analytics and consent can use this extension and wait for it to write the sync'd CID to cookies. Additionally, `amp-consent` would then also have the Linker feature and we could move away from solely depending on `amp-analytics` to handle CID syncing (since consent needs it too).

For now, to fix this problem immediately, let's fork an abbreviated version of the CookieWriter feature from `amp-analytics` into `amp-consent`. We will use the same config for the consent cookie writer that exists for the [analytics cookie writer](https://github.com/ampproject/amphtml/blob/master/extensions/amp-analytics/linker-id-receiving.md#receiving-linker-params-on-amp-pages).

```
'cookies': {
  'enabled': true,
  'cookieName1': {
    'value': 'QUERY_PARAM(example)',
  },
  'cookieName2': {
    'value': 'LINKER_PARAM(exampleParamName, exampleIdName)',
  },
}
``` 

**Intricacies**

We will only, allow the LINKER_PARAM macro to be used in the CookieWriter, and we will not clean the URL params (since we don't want a race with amp-analytics LINKER). If the url needs to be cleaned, then amp-analytics Linker will still need to be used (if it occurs first then we don't have a problem). No sameSite capabilities, as we are only using this for CID syncing, and not to be used by 3p iframe.

Let's check the binary size change as well: 1.87kb for `.mjs` and `.js`

### CODE REVIEW

File: extensions/amp-consent/0.1/amp-consent.js [hunk 1]
```diff
@@ -220,10 +221,17 @@ export class AmpConsent extends AMP.BaseElement {
       this.notificationUiManager_ = /** @type {!NotificationUiManager} */ (manager);
     });
 
+    const configRewritterPromise = new CookieWriter(
```
Comment (dmanek) [inline extensions/amp-consent/0.1/amp-consent.js:224]: `configRewritterPromise` -> `configRewriterPromise` (extra `t`)

File: extensions/amp-consent/0.1/cookie-writer.js [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+    }
+
+    if (!isObject(this.config_['cookies'])) {
+      user().error(TAG, 'cookies config must be an object');
+      return Promise.resolve();
+    }
+
+    const inputConfig = this.config_['cookies'];
+
+    if (inputConfig['enabled'] === false) {
```
Comment (dmanek) [inline extensions/amp-consent/0.1/cookie-writer.js:110]: Can this be a falsy value? If so change to `if (!inputConfig['enabled'])`

File: extensions/amp-consent/0.1/cookie-writer.js [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+   * value is found or value is invalid
+   * @param {JsonObject} inputConfig
+   * @return {number}
+   */
+  getCookieMaxAgeMs_(inputConfig) {
+    if (!hasOwn(inputConfig, 'cookieMaxAge')) {
+      return BASE_CID_MAX_AGE_MILLIS;
+    }
+
+    const cookieMaxAgeNumber = Number(inputConfig['cookieMaxAge']);
```
Comment (dmanek) [inline extensions/amp-consent/0.1/cookie-writer.js:147]: Drop Number from `cookieMaxAgeNumber`

**Review by dmanek — COMMENTED** [review]
There's a lot of encoding, serializing, etc. going on in this PR. Is there an npm package we can use instead?

**Comment by micajuine-ho** [comment]
> There's a lot of encoding, serializing, etc. going on in this PR. Is there an npm package we can use instead?

Unfortunately, we have to keep all of this encoding, serializing, etc, the same as the CW in `amp-analytics`. Additionally, changing this on them would be a breaking change to publishers who decode the LinkerParam on the origin, not using AMP.

File: extensions/amp-consent/0.1/cookie-writer.js (hunk 2 above)
Comment (micajuine-ho) [reply extensions/amp-consent/0.1/cookie-writer.js:110]: `enabled` is true by default, so devs have to explicitly set `enabled` to false.

**Review by dmanek — APPROVED** [review]

**Review by estherkim — APPROVED** [review]

**Pull request merged**